### PR TITLE
feat(api): PartialUpdateHostTemplate endpoint: handle macros

### PR DESCRIPTION
**Pull request by a-launois**
## Description

New endpoint for host template configuration PartialUpdateHostTemplate
```
PATCH /configuration/hosts/template/<hostTemplateId>
```
Note : this endpoint only handle property macros at the moment, more properties will be added gradually

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/doc/API/v23.10/Configuration/HostTemplate/HostTemplate.yaml [hunk 1]
```diff
@@ -1,24 +1,47 @@
...
+    '500':
+      $ref: '../../Common/Response/InternalServerError.yaml'
+patch:
+  tags:
+    - Host template
+  summary: "Patch a host template configuration"
+  description: |
+    Patch a host template configuration.
+
+    This endpoint do no handle all host template properties.
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/HostTemplate/HostTemplate.yaml:25]:
```suggestion
    This endpoint does not handle all host template properties.
```

File: centreon/doc/API/v23.10/Configuration/HostTemplate/Schema/PatchHostTemplateRequest.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+type: object
+required: []
+properties:
+  macros:
+    type: array
+    description: |
+      Host macros defined for the host (directly or through template or command inheritance)
```
Comment (cg-tw) [inline centreon/doc/API/v23.10/Configuration/HostTemplate/Schema/PatchHostTemplateRequest.yaml:7]:
```suggestion
      Host macros defined for the host (directly or through a template or command inheritance)
```

File: centreon/src/Core/CommandMacro/Infrastructure/Repository/DbReadCommandMacroRepository.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        $this->db = $db;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function findByCommandIdAndType(int $commandId, CommandMacroType $type): array
+    {
+        $this->info('Get command macros by command id and type',['command_id' => $commandId, 'type' => $type]);
```
Comment (cg-tw) [inline centreon/src/Core/CommandMacro/Infrastructure/Repository/DbReadCommandMacroRepository.php:50]:
```suggestion
        $this->info('Get command macros by command ID and type',['command_id' => $commandId, 'type' => $type]);
```

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php [hunk 4]
```diff
@@ -308,6 +308,36 @@ public function findById(int $hostTemplateId): ?HostTemplate
         return $this->createHostTemplateFromArray($result);
     }
 
+    /**
+     * @inheritDoc
+     */
+    public function findInheritanceLine(int $hostTemplateId): array
+    {
+        $this->info('Find parents ids of host template with id #' . $hostTemplateId);
```
Comment (cg-tw) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php:316]:
```suggestion
        $this->info('Find parents IDs of host template with ID #' . $hostTemplateId);
```

File: centreon/src/Core/HostMacro/Application/Repository/WriteHostMacroRepositoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+     * Add host's macros.
+     *
+     * @param HostMacro $macro
+     *
+     * @throws \Throwable
+     */
+    public function add(HostMacro $macro): void;
+
+    /**
+     * Add host's macros.
```
Comment (callapa) [inline centreon/src/Core/HostMacro/Application/Repository/WriteHostMacroRepositoryInterface.php:40]:
```suggestion
     * Update host's macros.
```

File: centreon/src/Core/HostMacro/Domain/Model/HostMacro.php [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    public function getDescription(): string
+    {
+        return $this->description;
+    }
+
+    public function setDescription(string $description): void
+    {
+        $shortName = (new \ReflectionClass($this))->getShortName();
```
Comment (callapa) [inline centreon/src/Core/HostMacro/Domain/Model/HostMacro.php:96]: You can use the $shortName variable defined in the constructor by using a private property.

File: centreon/src/Core/HostMacro/Domain/Model/HostMacroDifference.php [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    /** @var array<string,HostMacro> */
+    private array $commonMacros = [];
+
+    /** @var array<string,HostMacro> */
+    private array $updatedMacros = [];
+
+    /**
+     * Compare $afterMacros to $directMacros, $inheritedHostMacros and $inheritedCommandMacros
+     * and determine with ones had been added, updated, deleted.
+     *
```
Comment (callapa) [inline centreon/src/Core/HostMacro/Domain/Model/HostMacroDifference.php:44]: Move the description to the class header.

File: centreon/src/Core/HostTemplate/Application/Exception/HostTemplateException.php [hunk 8]
```diff
@@ -51,6 +51,14 @@ public static function addHostTemplate(): self
         return new self(_('Error while adding a host template'));
     }
 
+    /**
+     * @return self
+     */
+    public static function patchHostTemplate(): self
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/Exception/HostTemplateException.php:57]: not patch

File: centreon/src/Core/HostTemplate/Application/UseCase/PatchHostTemplate/PatchHostTemplate.php [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+use Core\CommandMacro\Domain\Model\CommandMacroType;
+use Core\HostMacro\Application\Repository\ReadHostMacroRepositoryInterface;
+use Core\HostMacro\Application\Repository\WriteHostMacroRepositoryInterface;
+use Core\HostMacro\Domain\Model\HostMacro;
+use Core\HostMacro\Domain\Model\HostMacroDifference;
+use Core\HostTemplate\Application\Exception\HostTemplateException;
+use Core\HostTemplate\Application\Repository\ReadHostTemplateRepositoryInterface;
+use Core\HostTemplate\Domain\Model\HostTemplate;
+
+final class PatchHostTemplate
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/PatchHostTemplate/PatchHostTemplate.php:47]:
```suggestion
final class PartialUpdateHostTemplate
```

**Review by cg-tw — APPROVED** [review]

File: centreon/src/Core/HostMacro/Domain/Model/HostMacroDifference.php [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+namespace Core\HostMacro\Domain\Model;
+use Core\CommandMacro\Domain\Model\CommandMacro;
+
+/**
+ * Compare $afterMacros to $directMacros, $inheritedHostMacros and $inheritedCommandMacros
+ * to determine witch afterMacros were added, updated, deleted or are unchanged.
+ *
+ * Require and return arrays with macro names as keys.
+ *
+ * TODO : link to confluence doc ?
```
Comment (a-launois) [inline centreon/src/Core/HostMacro/Domain/Model/HostMacroDifference.php:33]:
```suggestion
```

**Review by jeremyjaouen — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**